File: selenium-agent-mcp-starter/pages/inventory.page.ts
import { By } from 'selenium-webdriver';
import { BasePage } from './base.page';

export class InventoryPage extends BasePage {
  private title = By.css('.title');
  private cartLink = By.css('.shopping_cart_link');
  private cartBadge = By.css('.shopping_cart_badge');
  private inventoryItems = By.css('.inventory_item');
  private sortSelect = By.css('[data-test="product-sort-container"]');
  private menuButton = By.id('react-burger-menu-btn');
  private logoutLink = By.id('logout_sidebar_link');
  private aboutLink = By.id('about_sidebar_link');
  private resetAppStateLink = By.id('reset_sidebar_link');
  private firstItemName = By.id('item_4_title_link');
  private firstAddToCartButton = By.id('add-to-cart-sauce-labs-backpack');
  private firstRemoveButton = By.id('remove-sauce-labs-backpack');
  private secondAddToCartButton = By.id('add-to-cart-sauce-labs-bike-light');
  private itemNames = By.css('.inventory_item_name');

  async getTitleText(): Promise<string> {
    return this.getText(this.title);
  }

  async getItemCount(): Promise<number> {
    const items = await this.getElements(this.inventoryItems);
    return items.length;
  }

  async openFirstItem(): Promise<void> {
    await this.jsClick(this.firstItemName);
    await this.waitForUrlContains('inventory-item.html');
  }

  async addFirstItemToCart(): Promise<void> {
    await this.jsClick(this.firstAddToCartButton);
    await this.driver.wait(async () => {
      if (await this.isCartBadgeVisible()) {
        return true;
      }
      return this.isVisible(this.firstRemoveButton);
    }, 5000);
  }

  async removeFirstItemFromCart(): Promise<void> {
    await this.jsClick(this.firstRemoveButton);
    await this.driver.wait(async () => !(await this.isCartBadgeVisible()), 5000);
  }

  async openCart(): Promise<void> {
    const currentUrl = await this.getCurrentUrl();
    const cartUrl = new URL('cart.html', currentUrl).toString();

    try {
      await this.jsClick(this.cartLink);
      await this.waitForUrlContains('cart.html');
    } catch {
      await this.navigate(cartUrl);
      await this.waitForUrlContains('cart.html');
    }
  }

  async getCartBadgeText(): Promise<string> {
    return this.getText(this.cartBadge);
  }

  async isCartBadgeVisible(): Promise<boolean> {
    return this.isVisible(this.cartBadge);
  }

  async getCurrentSortValue(): Promise<string> {
    return this.getAttribute(this.sortSelect, 'value');
  }

  async sortBy(value: string): Promise<void> {
    const select = await this.waitForVisible(this.sortSelect);
    await this.driver.executeScript(
      'arguments[0].value = arguments[1]; arguments[0].dispatchEvent(new Event("change", { bubbles: true }));',
      select,
      value,
    );
    await this.driver.wait(async () => (await this.getCurrentSortValue()) === value, 5000);
  }

  async getItemNames(): Promise<string[]> {
    const elements = await this.getElements(this.itemNames);
    const names = await Promise.all(elements.map((element) => element.getText()));
    return names.filter(Boolean);
  }

  async addSecondItemToCart(): Promise<void> {
    await this.jsClick(this.secondAddToCartButton);
    await this.driver.wait(async () => (await this.getCartBadgeText()) === '2', 5000);
  }

  async openMenu(): Promise<void> {
    await this.jsClick(this.menuButton);
    await this.waitForVisible(this.logoutLink);
  }

  async clickLogout(): Promise<void> {
    await this.openMenu();
    await this.jsClick(this.logoutLink);
    await this.waitForUrlContains('saucedemo.com');
  }

  async clickAbout(): Promise<void> {
    await this.openMenu();
    await this.jsClick(this.aboutLink);
    await this.driver.wait(async () => (await this.getCurrentUrl()).includes('saucelabs.com'), 10000);
  }

  async clickResetAppState(): Promise<void> {
    await this.openMenu();
    await this.jsClick(this.resetAppStateLink);
    await this.driver.wait(async () => !(await this.isCartBadgeVisible()), 5000);
  }
}

File: selenium-agent-mcp-starter-python/src/pages/inventory_page.py
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import Select
from src.pages.base_page import BasePage


class InventoryPage(BasePage):
    _title = (By.CSS_SELECTOR, ".title")
    _app_logo = (By.CSS_SELECTOR, ".app_logo")
    _inventory_items = (By.CSS_SELECTOR, ".inventory_item")
    _sort_select = (By.CSS_SELECTOR, ".product_sort_container, [data-test='product-sort-container']")
    _cart_link = (By.CSS_SELECTOR, ".shopping_cart_link, [data-test='shopping-cart-link']")
    _cart_badge = (By.CSS_SELECTOR, ".shopping_cart_badge, [data-test='shopping-cart-badge']")
    _menu_button = (By.ID, "react-burger-menu-btn")
    _logout_link = (By.ID, "logout_sidebar_link")
    _about_link = (By.ID, "about_sidebar_link")
    _backpack_item_name = (By.ID, "item_4_title_link")
    _backpack_add_button = (By.CSS_SELECTOR, "#add-to-cart-sauce-labs-backpack, [data-test='add-to-cart-sauce-labs-backpack']")
    _backpack_remove_button = (By.CSS_SELECTOR, "#remove-sauce-labs-backpack, [data-test='remove-sauce-labs-backpack']")

    def is_loaded(self) -> bool:
        return self.is_visible(self._title) and self.get_text(self._title) == "Products"

    def get_logo_text(self) -> str:
        return self.get_text(self._app_logo)

    def get_inventory_count(self) -> int:
        return len(self.find_elements(self._inventory_items))

    def get_selected_sort(self) -> str:
        return Select(self.find_element(self._sort_select)).first_selected_option.text

    def add_backpack_to_cart(self):
        self.js_click(self._backpack_add_button)

    def remove_backpack_from_cart(self):
        self.js_click(self._backpack_remove_button)

    def open_backpack_details(self):
        self.js_click(self._backpack_item_name)
        self.wait_for_url_contains("inventory-item.html")

    def open_cart(self):
        self.js_click(self._cart_link)
        self.wait_for_url_contains("cart.html")

    def get_cart_badge_count(self) -> str:
        return self.get_text(self._cart_badge)

    def is_backpack_remove_visible(self) -> bool:
        return self.is_visible(self._backpack_remove_button)

    def open_menu(self):
        self.js_click(self._menu_button)
        self.wait_for_visible_element(self._logout_link)

    def logout(self):
        self.open_menu()
        self.js_click(self._logout_link)
        self.wait_for_url_contains("saucedemo.com")

    def open_about(self):
        self.open_menu()
        self.click(self._about_link)
